### ISSUE TITLE
SmartAPI data does not prescribe a ServiceID

### ISSUE THREAD

**Issue by jeffhhk** [SmartAPI/smartAPI#136]
Data collected by SmartAPI appears not to prescribe an official Service ID.

While Translator services implement a common API (TRAPI), services differ in scope of concern, and services sometimes implement specialized functionality.

Let's define "ServiceID" to be a field for identifying a service, invariant of its version and ci/dev/prod environment.  As far as I can tell, SmartAPI does not prescribe use of a ServiceID field.

Because there is no official ServiceID field, humans manually maintain various incompatible notions of "ServiceID".  Three examples of de facto ServiceIDs:

1) "Translator Environment" spreadsheet:
> https://docs.google.com/spreadsheets/d/1Qge8SXmPhiI8xM_gBWprlWJKL_hlPEiLyefJdXJ72II/edit#gid=0
    The column "Tool" is a de facto ServiceID.

2) For ARAX to call other services, it uses two "endpoint_map" dictionaries.  The key of each dictionary is a de facto ServiceID.
> https://github.com/RTXteam/RTX/blob/e4da9fe2d1d6d2450febf9bbb124421128cb27e7/code/ARAX/ARAXQuery/Expand/expand_utilities.py#L525-L561

3) In the ARS, each call of the ARSAppConfig base class constructor passes app_path, a de facto ServiceID.
> https://github.com/NCATSTranslator/Relay/blob/837ca86dd3b813fe2d5c5c08ee6d135b23717107/tr_sys/tr_ara_bte/bte_app.py#L5-L8

As an example of why we might want a ServiceID, consider the Translator Architecture WG 2021-10-19 presentation "Dev v Demo" https://docs.google.com/presentation/d/1VgGD3sFQkfseFAsJVpe3ZD8DmmmSKfAe8Kwc06Y7BYE/edit#slide=id.gf90803759e_0_63

The presentation's conclusion is to propose a change:

> The Ask, by Nov 9:
> 
>  Each team has 2 instances of their tool.
>  Both instances have a smartAPI registration.
>  One instance (the one labeled “dev”) can change freely, as we currently do
>  One instance (the one labeled “demo”) must remain fixed unless change approved by the group of workflow owners (?)
>  Each tool must configure itself to only talk to components in the same environment as (dev/demo) as it is in.

If we had a ServiceID in smart-api, the spreadsheet above, the ARS and the ARAX service could all be updated without manual human action, and with significantly lower meeting/coordination overhead.

Let's consider the existing smart-api fields and show why they are not being treated as ServiceIDs:

Is .info.x-translator.component a ServiceID?  No, because it holds "KP" for every KP, not a value describing which KP.

Is .info.title a ServiceID?  While some services seem to use the .info.title field in a way consistent with being treated as an ID, for example:

    curl -X GET 'https://smart-api.info/api/query?q=tags.name:translator&size=150&fields=_meta,info,servers' --silent | jq -c '.hits[] | select(.info."x-trapi".version != null)' | grep maturity | head -1 | jq .
    {
      "_id": "dc91716f44207d2e1287c727f281d339",
      "_meta": {
        "date_created": "2021-03-11T18:52:08.763346+00:00",
        "last_updated": "2021-10-21T07:01:50.636960+00:00",
        "url": "https://raw.githubusercontent.com/NCATS-Tangerine/translator-api-registry/master/biothings_explorer/smartapi.yaml",
        "username": "andrewsu"
      },
      "_score": 0.42180508,
      "info": {
        "contact": {
          "email": "asu@scripps.edu",
          "name": "Andrew Su",
          "url": "https://github.com/andrewsu"
        },
        "description": "A ReasonerStdAPI for BioThings Explorer",
        "termsOfService": "https://biothings.io/about",
        "title": "BioThings Explorer ReasonerStdAPI",
        "version": "2.8.1",
        "x-translator": {
          "biolink-version": "2.2.5",
          "component": "ARA",
          "team": [
            "Exploring Agent"
          ]
        },
        "x-trapi": {
          "asyncquery": true,
          "operations": [
            "lookup"
          ],
          "version": "1.2.1"
        }
      },
      "servers": [
        {
          "description": "Production server",
          "url": "https://api.bte.ncats.io/v1",
          "x-maturity": "production"
        },
        {
          "description": "Dev server",
          "url": "https://dev.api.bte.ncats.io/v1/",
          "x-maturity": "development"
        },
        {
          "description": "CI Dev server",
          "url": "https://bte.ci.transltr.io/v1/",
          "x-maturity": "development-ci"
        }
      ]
    }

other services include information such as TRAPI version in their title, making title not invariant of version:
    
    curl -X GET 'https://smart-api.info/api/query?q=tags.name:translator&size=150&fields=_meta,info,servers' --silent | jq -c '.hits[] | select(.info."x-trapi".version != null)' | grep https://automat.renci.org/textminingkp/1.2 | head -1 | jq .
    {
      "_id": "d2eecc42b6ae9109b12bcd215c787479",
      "_meta": {
        "date_created": "2021-08-30T19:08:59.380379+00:00",
        "last_updated": "2021-10-21T07:03:57.698024+00:00",
        "url": "https://automat.renci.org/textminingkp/1.2/openapi.json",
        "username": "YaphetKG"
      },
      "_score": 0.5768626,
      "info": {
        "contact": {
          "email": "kebedey@renci.org",
          "name": "Yaphet Kebede",
          "x-id": "https://github.com/yaphetkg",
          "x-role": "contributor"
        },
        "description": " A literature co-occurrence graph created by the Translator Text Mining KP.",
        "termsOfService": "http://robokop.renci.org:7055/tos?service_long=Automat Textmining KP&provider_long=RENCI",
        "title": "Automat Textmining KP (trapi v-1.2.0)",
        "version": "1.1.3",
        "x-translator": {
          "biolink-version": "1.8.2",
          "component": "KP",
          "team": [
            "Ranking Agent"
          ]
        },
        "x-trapi": {
          "operation": [
            "lookup",
            "overlay_connect_knodes",
            "annotate_nodes"
          ],
          "version": "1.2.0"
        }
      },
      "servers": [
        {
          "description": "Default server",
          "url": "https://automat.renci.org/textminingkp/1.2"
        }
      ]
    }

Is .info.x-translator.team a ServiceID?  No, because some teams have multiple services.

In conclusion, there is no existing ServiceID.  But there should be.
